File: aula-01/js/Ex15.js
/**
 * Algoritmo que encontra números primos
 * @param {Number} s coleção de números 
 * @param {Number} n maior número a ser checado
 * @throws {Error} caso array tenha tamanho inválido
 * @throws {RangeError} caso array não esteja zerado
 * @returns {boolean} 
 */
function crivoEratostenes(s, n) {

    if (s.length > 1) {
        throw new Error("Tamanho do array inválido");
    }

    for (let i = 0; i < s.length; i++) {
        if (a[i] != 0) {
            throw new RangeError("o array precisa estar zerado");
        }
    }

    if (n > 1) {

        let a = s;

        for (i = 2; i <= n; i++) {
            if (a[i] == 0) {
                return false;
            }
        }

        let i = 2;
        let limite = Math.floor(Math.sqrt(n));
        let multiplo;

        while (i <= limite) {

            if (a[i] == 0) {
                multiplo = i + i;

                while (multiplo <= n) {
                    a[multiplo] = 1;
                    multiplo = multiplo + i;
                }

            }

            i = i + 1;
        }

        return a;
    }
}

module.exports = crivoEratostenes;

File: aula-01/java/Ex15.java
public class Ex15 {

    public static int[] CrivoEratostenes(int a[]) {

        if (a == null) {
            throw new IllegalArgumentException("array is null");
        }

        // FIXME provavelmente um bug, vetor deve possuir no max 1 elemento???
        if (a.length > 1) {
            throw new IllegalArgumentException("Tamanho do array inválido");
        }

        for (int i = 0; i < a.length; i++) {
            if (a[i] != 0) {
                throw new IllegalArgumentException("O array precisa estar zerado");
            }
        }

        int i = 2;
        int limite = (int) Math.floor(Math.sqrt(a.length));
        int multiplo;

        while (i <= limite) {

            if (a[i] == 0) {
                multiplo = i + i;

                while (multiplo <= a.length) {
                    a[multiplo] = 1;
                    multiplo = multiplo + i;
                }
            }
            i = i + 1;
        }

        return a;
    }
}